Update main and registry [part 3]

r0mainK opened the pull request:
Now that new index is implemented I update the registry and main to use it. Also added a small error to the backend but it is not relevant (yet) because changes to the backend will be in next part. I also commented for now the initialize_registry backend method for the same reason. Models are slightly changed to not use the index from the backend.

vmarkovtsev commented on modelforge/__main__.py:
```diff
@@ -24,39 +23,67 @@ def add_backend_args(p):
 
+    def add_index_args(p, required=False):
+        p.add_argument("--username", default="", required=required,
+                       help="Username for git based index.")
```
for the Git repository with the index

vmarkovtsev commented on modelforge/__main__.py:
```diff
@@ -24,39 +23,67 @@ def add_backend_args(p):
+                       help="Username for git based index.")
+        p.add_argument("--password", default="", required=required,
+                       help="Password for git based index.")
+        p.add_argument("--remote_repo", default="", help="Git repository to use for index.")
```
`--remote`

The URL of the Git repository with the index.

r0mainK replied:
I think i will use three flags here: `--remote`, `--protocol` (multi choice, defaults to https) and `--domain` (defaults to github.com). This will be easier, if we simply ask for the url there is a good chance parsing it will be cumbersome for custom git platforms

vmarkovtsev replied:
OK, let's not reinvent the wheel and consume the git remote specification format

vmarkovtsev commented on modelforge/__main__.py:
```diff
@@ -24,39 +23,67 @@ def add_backend_args(p):
-    add_backend_args(publish_parser)
+    publish_parser.add_argument(
+        "--meta", default=os.path.join(os.path.dirname(__file__), "templates/template_meta.json"),
+        help="Path to a JSON containing extra metadata for the model.")
```
Path to the JSON file which contains the additional metadata of the model

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -29,36 +31,46 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
+        model_url = backend.upload_model(path, base_meta, args.force)
+    except ModelExistsError:
+        return 1
+    with open(os.path.join(args.meta), "r") as _in:
```
`"r"` can be deleted

also `encoding="utf-8"`

r0mainK commented:
@vmarkovtsev i think you can review, I will add more tests for coverage in next PRs if needed

vmarkovtsev commented on modelforge/__main__.py:
```diff
@@ -24,39 +25,73 @@ def add_backend_args(p):
+                       help="Protocol to use to connect to the Git repository.")
+        p.add_argument("--domain", default=DEFAULT_DOMAIN,
+                       help="Domain where the Git repository lives.")
+        p.add_argument("--repo", default=DEFAULT_REPO, help="Git repository to use for index.")
```
So we want to write `https://github.com/src-d/models` or `ssh://git@github.com/src-d/models.git`and do not parse URLs by hand. Try this: https://docs.python.org/3/library/urllib.parse.html

r0mainK replied:
okay, will replace with single `url` flag and parse in init of Index

vmarkovtsev commented on modelforge/__main__.py:
```diff
@@ -24,39 +25,73 @@ def add_backend_args(p):
     publish_parser = subparsers.add_parser(
-        "publish", help="Upload the model to the cloud and update the "
-                        "registry.")
+        "publish", help="Upload the model to the cloud and update the registry.")
```
Not necessarily cloud. The backend can be anything, even `/dev/null`

r0mainK replied:
removing

vmarkovtsev commented on modelforge/model.py:
```diff
@@ -42,14 +41,14 @@ def __init__(self, **kwargs):
-    def load(self, source: Union[str, BinaryIO, "Model"]=None,
-             cache_dir: str=None, backend: StorageBackend=None) -> "Model":
+    def load(self, source: Union[str, BinaryIO, "Model"]=None, cache_dir: str=None,
+             backend: StorageBackend=None, index: dict=None) -> "Model":
```
`load()` should be as smooth API as possible, so better to revert to the old behavior. The index should be retrieved automatically.

r0mainK replied:
are you sure ? I had to change this because we would fetch the index from the backend before, but since it is now git index we can't do that. I think passing this dict was the easiest, but we could also extract the model_type beforehand and pass it directly, since we do not really need the dict.

vmarkovtsev commented on modelforge/model.py:
```diff
@@ -260,7 +258,6 @@ def save(self, output: Union[str, BinaryIO], deps: Iterable=tuple(),
              create_missing_dirs: bool=True) -> "Model":
         """
         Serializes the model to a file.
-
```
You need to ensure that the removal of this newline does not break the docstrings.

r0mainK replied:
Shouldnt, this is just to homogenize with other doctstrings in the file

vmarkovtsev commented on modelforge/models.py:
```diff
@@ -26,12 +25,12 @@ class GenericModel(Model):
-    def __init__(self, source: Union[str, "Model"]=None, dummy=False,
-                 cache_dir: str=None, backend: StorageBackend=None,
+    def __init__(self, source: Union[str, "Model"]=None, dummy=False, cache_dir: str=None,
+                 backend: StorageBackend=None, index: dict=None,
```
same here

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -29,36 +31,47 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
+                         username=args.username, password=args.password, cache=args.cache,
+                         log_level=args.log_level)
+    template_model = git_index.load_template(args.template_model)
+    git_index.add_model(base_meta["model"], base_meta["uuid"], base_meta, extra_meta, model_url,
```
`meta` should be single. Merge should be performed beforehand.

r0mainK replied:
on it, gonna change the templates anyway since we also change description/code

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -29,36 +31,47 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
+    git_index = GitIndex(protocol=args.protocol, domain=args.domain, repo=args.repo,
+                         username=args.username, password=args.password, cache=args.cache,
+                         log_level=args.log_level)
+    for key, val in git_index.index["models"].items():
```
It should be `.models` property

r0mainK replied:
done

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -29,36 +31,47 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
                 break
         for mid, meta in sorted(
-                [m for m in val.items() if m[1] != default],
+                [m for m in val.items() if m[1] != default and
```
code and description must be a top-level constant;

besides, better to add a new var because this has grown too big

r0mainK replied:
ok, so atm the index.json has this form:
```
{"models":{"model_type_1":{"code": "...", "description": ..., "default": "uuid",  "uuid_1": { ... }, "uuid_2": { ... }}, "model_type_2": { ... } }, "datasets": { ... }}
```
Note that the `datasets` is currently not implemented / used. Anyway I think we can change to this:
```
{"models":{"model_type_1":{"uuid_1": { ... }, "uuid_2": { ... }}, "model_type_2": { ... } }, 
"meta": {"model_type_1": {"code": "", "description": "", "default": "uuid"}, "model_type_2": { ... } } }
```
The metadata specific to each model will still be kept in the dict in the appropriate model_type/ uuid, but the rest like code or description will be stored here.

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -68,19 +81,58 @@ def list_models(args: argparse.Namespace, backend: StorageBackend, log: logging.
+    :return: None
+    """
+
+    git_index = GitIndex(protocol=args.protocol, domain=args.domain, repo=args.repo,
```
Makes sense to add `@supply_index` decorator and remove this boilerplate

r0mainK replied:
on it

vmarkovtsev commented on modelforge/gcs_backend.py:
```diff
@@ -49,7 +48,8 @@ class GCSBackend(StorageBackend):
     DEFAULT_CHUNK_SIZE = 65536
     INDEX_FILE = "index.json"
```
I guess this should be removed too

vmarkovtsev commented on modelforge/gcs_backend.py:
```diff
@@ -78,11 +79,7 @@ def fetch_model(self, source: str, file: str) -> None:
         self._fetch(source, file)
 
     def fetch_index(self) -> dict:
```
Better to make it a property

```python
@property
def index(self):
    return self._index
```

vmarkovtsev commented on modelforge/index.py:
```diff
@@ -25,38 +23,46 @@
-        "initilialize": "Initialize a new Modelforge index",
-        "delete": "Delete {model_type}/{model_uuid}",
-        "add": "Add {model_type}/{model_uuid}",
+        "initilialize": b"Initialize a new Modelforge index",
```
This is too specific, we should convert to bytes where it is needed

vmarkovtsev commented on modelforge/index.py:
```diff
@@ -26,37 +24,44 @@ class GitIndex:
-        :param protocol: Protocol to use, either http(s), tcp or ssh
-        :param domain: Domain name, defaults to github.com
-        :param repo: Remote repo to use for index, defaults to the one set by configuration.py file
+        :param url: Remote url to use for index
```
Remote repository's address where the index is maintained.

r0mainK replied:
done

vmarkovtsev commented on modelforge/index.py:
```diff
@@ -26,37 +24,44 @@ class GitIndex:
-        :raise ValueError: If missing credential, incorrect domain or incorrect credentials
+        :raise ValueError: If missing credential, incorrect url or incorrect credentials
         """
         self._log = logging.getLogger("GitIndex")
```
Even better: `getLogger(type(self).__name__)`

vmarkovtsev commented on modelforge/index.py:
```diff
@@ -26,37 +24,44 @@ class GitIndex:
+            raise ValueError("Parsed url does not contain a valid domain.")
+        if not parsed_url.path:
+            raise ValueError("Parsed url does not contain a valid repository path.")
+        self.repo = parsed_url.path.replace(".git", "")
```
only the ending should be truncated here
`mycrazy.git/repo.git`

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -1,23 +1,27 @@
+                        "update_default", "username", "password", "remote_repo", "template_model",
+                        "template_readme" and "log_level"
+    :param backend: Backend supplied to which publish the model
+    :param git_index: GitIndex supplied where the backend's index is located
```
Please change everywhere:

* Backend which is responsible for working with model files.
* GitIndex where the index is maintained.

r0mainK commented:
@vadim sorry this has become a bit huge -_- here is a quick recap:

- removed non relevant tests and changed a bunch of them -> still gotta add some, but for later PRs
- templates: had to change since we also change the form of the index
- index: in separate commit for now, mostly changes due to the previous review rounds
- main: nothing much changed since previous review rounds
- dump/delete: now in registry
- registry: now using the new index
- gcs_backend/backend: took out fetch_index abstract methods, replaced with property `index` that is defined in StorageBackend during init (thought it would be logical to always have git_index, whatever backend is being used) 
- model: replaced fetch_index with index property, also since the index changed had to change a tiny bit the logic
- meta: took out unused logic, changed a bit to gather all metadata from various sources for a model in this function

### INDEX CHANGES 

Before:

```
{"models":{
   "model_type_1": {
      "code": "readme_code", 
      "description": "readme_description", 
      "default": "uuid",  
      "uuid_1": { 
          "code": "model_code"
          "description": "model_description", ... }, 
      "uuid_2": { ... }},       
  "model_type_2": { ... } }, 
"datasets": { ... }}
```
Now:

```
{"models":{
  "model_type_1":{
     "uuid_1": { 
        "code": "model_code",
        "description": "model_description", ... } }, 
     "uuid_2": { ... }}, 
  "model_type_2": { ... } }, 
"meta": {
  "model_type_1": {"code": "readme_code", "readme_description": "", "default": "uuid"}, 
  "model_type_2": { ... } } }
```

As you can see we now separate the metadata used for the `readme.md` and `<uuid>.md`, the first is in the "meta" part, the rest is stored in the current entry in the "models" part.

vmarkovtsev commented on modelforge/index.py:
```diff
@@ -2,61 +2,63 @@
+        self.cached_repo = os.path.join(cache, self.repo)
+        if username and password:
+            auth = username + ":" + password + "@"
+            self.remote_url = REMOTE_URL % (parsed_url.scheme, auth, parsed_url.netloc, self.repo)
```
Since you only use username and password to form the `remote_url`, we could remove `--username` and `--password` completely and require them to be included into `--index-repo`

vmarkovtsev replied:
Ah, then we will lose the ability to configure the remote. All right.

vmarkovtsev commented on modelforge/registry.py:
```diff
@@ -1,23 +1,28 @@
 
 @supply_backend
-def publish_model(args: argparse.Namespace, backend: StorageBackend, log: logging.Logger):
+def publish_model(args: argparse.Namespace, backend: StorageBackend, git_index: GitIndex,
```
We can take the index from the backend. Thus `git_index` can be removed from the top level functions.

r0mainK replied:
True, I wanted to only put `git_index.index` in the backend since I saw no use of the rest, but yes we can do that. Changing

vmarkovtsev commented on modelforge/storage_backend.py:
```diff
@@ -14,14 +21,11 @@ def fetch_model(self, source: str, file: str) -> None:
-        """
-        raise NotImplementedError
+    @property
+    def git_index(self) -> GitIndex:
```
Better to name it just "index". Git is our implementation detail

vmarkovtsev commented on modelforge/backends.py:
```diff
@@ -50,13 +52,16 @@ def supply_backend_inner(func):
                 if backend is None:
                     return 1
-            return func(args, backend, log)
+            return func(args, backend, git_index, log)
```
I have an impression that we will break here because you removed it from the functions.

Hence the question: why did the tests pass?

r0mainK replied:
because I forgot to change the test for this function as I forgot to change it here, and I have not rewritten tests for the registry - yet. I took the old ones out because they were completely out of date.

vmarkovtsev commented:
@r0mainK Don't want to rape you any longer, just please rename `content` to `contents` in the next PR :)

r0mainK commented:
yeah no problem xDD
